text/0000-debug-improvements.md

```diff
@@ -0,0 +1,185 @@
+                    path: "/tmp/foobar",
+                    mode: R
+                },
+                buffer: 0/65536
+            }
+        ),
+        other_thing: 0
+    }
+}
+```
+However, we wouldn't want this "pretty printed" version to be used by default,
+since it's significantly more verbose.
+
+## Helper types
+
+For many Rust types, a Debug implementation can be automatically generated by
+`#[derive(Debug)]`. However, many encapsulated types cannot use the
+derived implementation. For example, the types in std::io::buffered all have
+manual `Debug` impls.  They all maintain a byte buffer that is both extremely
+large (64k by default) and full of uninitialized memory. Printing it in the
+`Debug` impl would be a terrible idea. Instead, the implementation prints the
+size of the buffer as well as how much data is in it at the moment:
+https://github.com/rust-lang/rust/blob/0aec4db1c09574da2f30e3844de6d252d79d4939/src/libstd/io/buffered.rs#L48-L60
+
+```rust
+pub struct BufferedStream<S> {
+    inner: BufferedReader<InternalBufferedWriter<S>>
+}
+
+impl<S> fmt::Debug for BufferedStream<S> where S: fmt::Debug {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        let reader = &self.inner;
+        let writer = &self.inner.inner.0;
+        write!(fmt, "BufferedStream {{ stream: {:?}, write_buffer: {}/{}, read_buffer: {}/{} }}",
+               writer.inner,
+               writer.pos, writer.buf.len(),
+               reader.cap - reader.pos, reader.buf.len())
+    }
+}
+```
+
+A purely manual implementation is tedious to write and error prone. These
+difficulties become even more pronounced with the introduction of the "pretty
+printed" format described above. If `Debug` is too painful to manually
+implement, developers of libraries will create poor implementations or omit
+them entirely. Some simple structures to help automatically create the correct
+output format can significantly help ease these implementations:
+```rust
+impl<S> fmt::Debug for BufferedStream<S> where S: fmt::Debug {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        let reader = &self.inner;
+        let writer = &self.inner.inner.0;
+        ShowStruct::new(fmt, "BufferedStream")
+            .field("stream", writer.inner)
+            .field("write_buffer", &format_args!("{}/{}", writer.pos, writer.buf.len()))
+            .field("read_buffer", &format_args!("{}/{}", reader.cap - reader.pos, reader.buf.len()))
+            .finish()
+    }
+}
+```
+
+# Detailed design
+
+## Pretty printing
+
+The `#` modifier (e.g. `{:#?}`) will be interpreted by `Debug` implementations
+as a request for "pretty printed" output:
+
+* Non-compound output is unchanged from normal `Debug` output: e.g. `10`,
+  `"hi"`, `None`.
+* Array, set and map output is printed with one element per line, indented four
+    spaces, and entries printed with the `#` modifier as well: e.g.
+```rust
+[
+    "a",
+    "b",
+    "c"
```

Comment:
I don't think the consistency is important, as I'd write the trailing comma in the pretty-printed style, but not in the compact one.